doc/src/explanation/best-practices.md

```diff
@@ -0,0 +1,1329 @@
+accept multiple forms of arguments, so long as it accepts the form stored in
+`args` (e.g., it is perfectly fine for some args to have default values).
+
+Here are some examples of reasons you might be tempted to override `__eq__`
+and the preferred alternatives:
+
+- To make `==` apply some smarter equality check than purely structural
+  equality. As noted above, this is a bad idea because too many things
+  implicitly assume `==` works structurally only. Instead, use a function or
+  method to implement the smarter equality checking (for example, the `equals`
+  method).
+
+  Another option is to define a {term}`canonicalization <canonicalize>` method
+  that puts objects into canonical form (e.g., via `doit`), so that, for
+  instance, `x.doit() == y.doit()` is true whenever `x` and `y` are
+  mathematically equal. This is not always possible because not every type of
+  object has a computable canonical form, but it is a convenient approach when
+  one does exist.
+
+- To make `==` check for some additional attributes beyond those stored in the
+  `args` of an expression. See the [](best-practices-extra-attributes) section
+  above for more details on why it's a bad idea to directly store extra
+  attributes on a SymPy object, and what the best alternatives are.
+
+- To make `==` compare equal to some non-SymPy object. It is preferable to
+  extend `sympify` to be able to convert this object into the SymPy object.
+  The default `__eq__` implementation will automatically call `sympify` on the
+  other argument if it isn't a `Basic` instance (e.g., `Integer(1) == int(1)`
+  gives `True`). It is possible to extend `sympify` both for objects you
+  control by defining a `_sympy_` method and for objects you do not control by
+  extending the `converter` dictionary. See the {func}`~.sympify`
+  documentation for more details.
+
+### Avoiding Infinite Recursion from Assumptions Handlers
+
+When writing assumptions handlers on custom functions like `_eval_is_positive`
+(see the [custom
+functions guide](custom-functions-assumptions) for details on how to do this),
+there are two important things to keep in mind:
+
+**Firstly, avoid creating new expressions inside of an assumption handler. You
+should always pull apart the arguments of a function directly instead.** The
+reason is that creating a new expression could itself result in an assumptions
+query. This can easily lead to infinite recursion. And even when it doesn't,
+creating a new expression that itself could lead to many recursive
+assumptions queries is bad for performance compared to querying the desired
+property more directly.
+
+This generally means using methods like {meth}`~.as_independent` or
+`{meth}`~.as_coeff_mul` and checking
+the `args` of expressions directly (see the [custom functions
+guide](custom-functions-assumptions) for an example).
+
+<!-- TODO: Give an example here. Can we show an infinite recursion? -->
```

Comment:
@oscarbenjamin do you know of an example of when assumptions handles can actually lead to infinite recursion? I couldn't come up with one easily, but I think maybe the problem is that this mainly happens inside of core SymPy classes rather than custom classes because those are the ones that are more likely to call each other.

Reply:
> do you know of an example of when assumptions handles can actually lead to infinite recursion?

You have an example below. Do you mean something from the codebase though?

Generally I would remove any code that does this from the codebase.

Reply:
I mean an illustrative example I can include here. The example below is based on infinite recursion from evaluating assumptions on `self`, but I am looking for an example based on the "don't create new expressions in assumptions handlers" rule.

Reply:
I can contrive an example but as you say it mostly happens in core class e.g. because someone does not consider that `1 + self` creates a new `Add` or something like that.

Here are examples of what should not be done:
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L618
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L661-L664
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L680
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L746
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L748-L750
https://github.com/sympy/sympy/blob/6daa2f06c61eeceb72690afe3055949682b22a49/sympy/core/power.py#L754-L756

There are many more in the same file. Maybe you can identify an immediate bug from those. There are two types of infinite recursion that can occur:

- Recursing onto a different object that then recurses onto another different object and so on.
- Recursing onto a larger object that contains the current object so that the query comes back down to the same query on the original object.

It is very difficult to predict when any created expressions might lead to either of these cases because automatic evaluation rules can be so unpredictable. I don't see how else to define a rule that could control it besides "never create a new expression".

Reply:
To be fair, some of those seem hard to do without just completely copying the assumptions logic from the other class. I guess I basically already said the same thing in the other issue, but this really is a fundamental limitation of this recursive way of implementing assumptions handlers, which makes them responsible for both returning the result of the assumptions query and, to some extent, the method by which other queries are logically deduced. A better design is to separate out the order that the queries are actually made in to a top-level solver, and represent facts in an symbolic way.

Anyway, I think this is probably hard to show an explicit example for. Maybe there's an issue or PR I could reference where something like this happened. Otherwise, I'm just going to leave this as it is and remove the TODO.